pubsub/google/cloud/pubsub_v1/publisher/client.py

```diff
@@ -268,21 +347,56 @@ def publish(self, topic, data, **attrs):
             )
 
         # Create the Pub/Sub message object.
-        message = types.PubsubMessage(data=data, attributes=attrs)
+        message = types.PubsubMessage(
+            data=data, ordering_key=ordering_key, attributes=attrs
+        )
 
-        # Delegate the publishing to the batch.
         with self._batch_lock:
             if self._is_stopped:
                 raise RuntimeError("Cannot publish on a stopped publisher.")
 
-            batch = self._batch(topic)
-            future = None
-            while future is None:
-                future = batch.publish(message)
-                if future is None:
-                    batch = self._batch(topic, create=True)
+            sequencer = self._get_or_create_sequencer(topic, ordering_key)
+
+            # Delegate the publishing to the sequencer.
+            future = sequencer.publish(message)
+
+            # Create a timer thread if necessary to enforce the batching
+            # timeout.
+            self._ensure_commit_timer_runs_no_lock()
+
+            return future
+
+    def ensure_commit_timer_runs(self):
+        """ Ensures a commit timer thread is running. If a commit timer thread
+            is already running, it does nothing.
+        """
+        with self._batch_lock:
+            self._ensure_commit_timer_runs_no_lock()
+
+    def _ensure_commit_timer_runs_no_lock(self):
+        if not self._commit_thread and self.batch_settings.max_latency < float("inf"):
```

Comment:
Just as a precaution, let's add an explicit comment that this method assumes that its caller has acquired the `_batch_lock`.